src/sp_var_parser.h

```diff
@@ -0,0 +1,51 @@
+#ifndef SP_VAR_PARSER_H
+# define SP_VAR_PARSER_H
+# include "php_snuffleupagus.h"
+# include "sp_list.h"
+
+typedef enum {
+  OBJECT = 1,
+  ARRAY,
+  ARRAY_END,
+  STRING_DELIMITER,
+  CLASS,
+  VAR,
+  ESC_STRING_DELIMITER,
+  CONSTANT
+} elem_type;
+
+typedef struct sp_token_s {
+  elem_type type;
+  char *token;
+  unsigned int pos;
+} sp_token_t;
+
+typedef struct parser_s {
+  elem_type type;
+  char *value;
+  struct parser_s *idx;
+  struct parser_s *next;
+} sp_tree;
+
+zval *get_value(zend_execute_data *, const sp_tree *, bool);
+sp_tree *sp_tree_new();
+sp_tree *parse_var(const char *);
+void print_type_list(const char *, sp_tree*, int);
+void sp_tree_free(sp_tree *);
+
+# define OBJECT_TOKEN "->"
+# define ARRAY_TOKEN "["
+# define ARRAY_END_TOKEN "]"
+# define STRING_TOKEN "\""
+# define ESC_STRING_TOKEN "\'"
+# define CLASS_TOKEN "::"
+
+# define VARIABLE_TOKEN '$'
+
+# define PRIVATE_PROP_FMT "%c%s%c%s"
+# define PROTECTED_PROP_FMT "%c*%c%s"
+
+# define REGEXP_VAR "^\\$[a-z_][a-z0-9_]*$"
```

Comment:
php supports non ascii variable names 
```
$ php -r '$räẗe = "4242";print($räẗe . "\n");'
```
but I guess your approach is safer for now :)

Reply:
We don't want people to use idiotic variable names in their rules. If a framework requires this kind of rules, I would argue that setting the physical box hosting it on fire is the recommended solution.

Reply:
php supports pretty much anything

```
$ php -r '${\'php support+@#!@!@$*(a)(!@$$!\|}{<> lot of things\'} = "qwe"; var_dump(get_defined_vars());'
```

Reply:
![sadness]( https://media.giphy.com/media/a9xhxAxaqOfQs/giphy.gif )